Library/Homebrew/test/test_cmd_audit.rb

```diff
@@ -0,0 +1,260 @@
+require "testing_env"
+require "fileutils"
+require "pathname"
+require "formulary"
+require "cmd/audit"
+
+class FormulaTextTests < Homebrew::TestCase
+  def setup
+    @dir = mktmpdir
+    # ensure we get a predictable environment
+    File.stubs(:umask).returns 022
+  end
+
+  def teardown
+    FileUtils.rm_rf @dir
+  end
+
+  def formula_text(name, text)
+      path = Pathname.new "#{@dir}/#{name}.rb"
+      path.open("w") { |f| f.write text }
+      FormulaText.new path
+  end
+
+  def test_simple_valid_formula
+    ft = formula_text "valid", <<-EOS.undent
+      class Valid < Formula
+        url "http://www.example.com/valid-1.0.tar.gz"
+      end
+    EOS
+
+    refute ft.has_DATA?, "The formula doesn't have DATA"
+    refute ft.has_END?, "The formula doesn't have __END__"
+    assert ft.has_trailing_newline?, "The formula have a trailing newline"
+
+    assert ft =~ /\burl\b/, "The formula match 'url'"
+    assert_nil ft.line_number(/desc/), "The formula doesn't match 'desc'"
+    assert_equal 2, ft.line_number(/\burl\b/)
+  end
+
+  def test_trailing_newline
+    ft = formula_text "newline", "class Newline < Formula; end"
+    refute ft.has_trailing_newline?, "The formula doesn't have a trailing newline"
+  end
+
+  def test_has_data
+    ft = formula_text "data", <<-EOS.undent
+      class Data < Formula
+        patch :DATA
+      end
+    EOS
+
+    assert ft.has_DATA?, "The formula has DATA"
+  end
+
+  def test_has_end
+    ft = formula_text "end", <<-EOS.undent
+      class End < Formula
+      end
+      __END__
+      a patch here
+    EOS
+
+    assert ft.has_END?, "The formula has __END__"
+    assert_equal "class End < Formula\nend", ft.without_patch
+  end
+end
+
+class FormulaAuditorTests < Homebrew::TestCase
+  def setup
+    @dir = mktmpdir
+  end
+
+  def teardown
+    FileUtils.rm_rf @dir
+  end
+
+  def formula_auditor(name, text, options = {})
+      path = Pathname.new "#{@dir}/#{name}.rb"
+      path.open("w") { |f| f.write text }
+      FormulaAuditor.new Formulary.factory(path), options
+  end
+
+  def test_init_no_problems
+    fa = formula_auditor "foo", <<-EOS.undent
+      class Foo < Formula
+        url "http://example.com/foo-1.0.tgz"
+      end
+    EOS
+
+    assert_equal [], fa.problems
+  end
+
+  def test_audit_file_permissions
+    fa = formula_auditor "foo", <<-EOS.undent
+      class Foo < Formula
+        url "http://example.com/foo-1.0.tgz"
+      end
+    EOS
+
+    path = fa.formula.path
+    path.chmod 0400
+
+    fa.audit_file
+    assert_equal ["Incorrect file permissions (400): chmod 644 #{path}"],
+      fa.problems
+  end
```

Comment:
This fails locally for me, because I don't have the default `umask` of `022`. The audit message I get is `Incorrect file permissions (400): chmod 600 <path>` and that doesn't match. Maybe you can make the match more tolerant?